### DISCUSSION TITLE
Add support SPI-UART adapters

### DISCUSSION THREAD

**Discussion by Seeung** [nanoframework/Home#1056] [Ideas]
### Description

I propose to add support for communication using SPI via UART.

### How to solve the problem

To do this, you need to implement `SpiDevice` over `Serial`. 

### Aditional context

An example of such a scenario is using `stc12f204` and `nrf24l01`.

**Comment by josesimoes** [comment]
This is not a bug. At most a feature request. Moving to discussions.

**Reply by Seeung** [reply]
I used `Feature request` template

**Comment by josesimoes** [comment]
@Seeung can you elaborate a bit more on the intended usage scenario?
You're mixing an 8 bit MCU with an RF module... that sounds a bit confusing and requires context.

**Reply by Seeung** [reply]
To use multiple `NRF24L01` modules on `ESP32` continuously, you need to use the appropriate number of `SPI interfaces`. In this situation, the solution for working with `NRF24L01` is implemented through the `UART` port using `STC12F204` which is a device for connecting `UART` and `SPI`.

**Reply by josesimoes** [reply]
I think I understand what you're describing. That seems like something that's application specific. 
Don't see a broader use of this that makes a justification of an official library for this...

**Reply by Seeung** [reply]
In the current version, I cannot use the official library to work with `NRF24L01` because it works with the `SpiDevice` class, if `SpiDevice` had a basic interface for implementing the layer, for example `ISpiDevice`, then this would simplify the work .

**Reply by Ellerbach** [reply]
I have to say I4m a bit confused as well... I see that the NFR24L01 you are mentioning is a SPI device: https://www.mouser.com/datasheet/2/297/nRF24L01_Product_Specification_v2_0-9199.pdf
So I don't understand what you want to do with UART as it doesn't seems to have UART at all.

**Reply by Seeung** [reply]
`STC12F204` adapter allows you to connect bus `UART` and `SPI`.

**Reply by Ellerbach** [reply]
So then what is your question? I still should be missing something, really sorry about that.

**Reply by Seeung** [reply]
The `NRF24L01` library uses the `SpiDevice` class to work, which only works with the `SPI` bus. In my scenario I am using `NRF24L01` communication via `STC12F204` adapter with which I can communicate with `NRF24L01` via `UART`. I would like to be able to use a ready-made solution or redefine `SpiDevice` without rebuilding libraries, for example, make the `SpiDevice` interface or an abstract class.

**Reply by alberk8** [reply]
So that means that on the nF side you should be using the UART library to talk to STC12F204.

**Reply by Seeung** [reply]
An example of using the adapter, I did not find it in English, only in Russian, there is an example of the implementation of the code for arduino.
https://kotvaska.medium.com/uart-%D0%B8-%D1%81%D0%B2%D1%8F%D0%B7%D1%8C-%D1%87%D0%B5%D1%80%D0%B5%D0%B7-%D0%BD%D0%B5%D0%B3%D0%BE-%D0%BF%D0%BE-nrf24l01-c5c68aafcc7b

**Reply by alberk8** [reply]
Like I say, you will be communicating with STC15FL204 via UART and that device will convert it to SPI for the NRF24L01.
